BUG: Don't cast categorical nan to int

dsaxton opened the pull request:
- [x] closes #28406
- [x] passes `black pandas`
- [x] tests added / passed
- [x] whatsnew entry

This raises an error when attempting to cast a `Categorical` or `CategoricalIndex` containing `nans` to an integer dtype.  Also had to remove the casting within `get_indexer_non_unique` since this won't always be possible.

Commit:
Don't cast categorical nan to int

jbrockmendel commented on pandas/tests/extension/test_categorical.py:
```diff
@@ -197,7 +197,14 @@ def test_searchsorted(self, data_for_sorting):
-    pass
+    def test_cast_nan_to_int(self):
+        s1 = pd.Series([0, 1, np.nan], dtype="category")
+        s2 = pd.Series([0, 1, np.inf], dtype="category")
```
might as well do -np.inf too.

Commit:
Parametrize test

Commit:
Add CategoricalIndex test

mroeschke commented on pandas/core/arrays/categorical.py:
```diff
@@ -520,6 +520,9 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
             if dtype == self.dtype:
                 return self
             return self._set_dtype(dtype)
+        if is_integer_dtype(dtype) and self.isin([np.nan, -np.inf, np.inf]).any():
```
`np.isfinite` might be better to use here.

dsaxton replied:
I'm going to revert back to using `isin` because `np.isfinite` breaks with input like `Series(["1", "2", "3"], dtype="category")`

TomAugspurger replied:
May also need a `is_float_dtype(self.categories.dtype)`

dsaxton replied:
Wouldn't we want it to raise regardless of the dtype of the categories?  E.g., `astype(int)` should raise for this object even though the categories are integers:
```
[1, 2, NaN]
Categories (2, int64): [1, 2]
```

dsaxton replied:
@TomAugspurger Do we also need `not is_extension_array_dtype(dtype)` here to rule out nullable integers?

TomAugspurger replied:
It looks like we follow this up with by passing `self` to `np.array`, which would lose the IntegerArray anyway.

Commit:
Use isfinite

Commit:
Fix

dsaxton commented:
FWIW the issue is here https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/base.py#L4717.  If `target` is `Categorical` then the values can't always be cast to the dtype of the categories.

jbrockmendel commented:
should the use_inf_as_na option matter here?

Commit:
Check TypeError as well for now

Categorical raises a ValueError at the moment, but CategoricalIndex
ends up raising a TypeError because this happens during the handling
of the ValueError

dsaxton commented:
> should the use_inf_as_na option matter here?

Wasn't familiar with that option, but it shouldn't affect `numpy` behavior I think?

gfyoung commented on pandas/tests/extension/test_categorical.py:
```diff
@@ -197,7 +197,13 @@ def test_searchsorted(self, data_for_sorting):
+    @pytest.mark.parametrize("cls", [Categorical, CategoricalIndex])
+    @pytest.mark.parametrize("value", [np.nan, -np.inf, np.inf])
+    def test_cast_nan_to_int(self, cls, value):
+        s = cls([0, 1, value])
```
Reference issue as a comment above.

gfyoung commented on pandas/tests/extension/test_categorical.py:
```diff
@@ -197,7 +197,13 @@ def test_searchsorted(self, data_for_sorting):
+    def test_cast_nan_to_int(self, cls, value):
+        s = cls([0, 1, value])
+
+        with pytest.raises((ValueError, TypeError)):
```
Check the error message too.

Commit:
Check error message

Commit:
Fix doc typo

Commit:
Revert "Use isfinite"

This reverts commit 504be9018e548d62937cce2466e5b2f1cb71f759.

Commit:
Extract array directly

Commit:
Keep _maybe_promote

Commit:
Add note

Commit:
Fix typo

WillAyd commented on pandas/core/indexes/base.py:
```diff
@@ -4713,13 +4713,13 @@ def set_value(self, arr, key, value):
 
-        if self.is_all_dates:
+        if is_categorical(target):
+            tgt_values = target.__array__()
```
@TomAugspurger thoughts on this approach?

TomAugspurger replied:
You shouldn't call `__array__` directly. You should use `np.asarray` instead if you need an ndarray.

TomAugspurger commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -110,6 +110,7 @@ Categorical
 ^^^^^^^^^^^
 
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
+- Bug in :meth:`astype` where ``nans`` were handled incorrectly when casting to int (:issue:`28406`)
```
```suggestion
- Bug in :meth:`Categorical.astype` where ``NaN`` values were handled incorrectly when casting to int (:issue:`28406`)
```

TomAugspurger reviewed (commented):
I think this PR should focus on just NaN handling.

I'd rather see `Float64Index` properly handle astyping inf values to int (raise), and then do `self.categories.take(self.codes)` with a check if any of the codes is negative, so that we raise.

dsaxton committed:
Update doc/source/whatsnew/v1.0.0.rst

Co-Authored-By: Tom Augspurger <TomAugspurger@users.noreply.github.com>

Commit:
Use np.asarray

TomAugspurger commented:
That would be my preference. Did my earlier comment make sense?

The `NaN` issue can be solved by checking `Categorical.codes`

```python
In [7]: c = pd.Categorical([1, None, 2])

In [8]: c.codes < 0
Out[8]: array([False,  True, False])
```

The inf issue is deeper, and is present in Float64Index, so I think it should be split to its own PR.

```python
In [9]: c = pd.Categorical([1, np.inf])

In [10]: c.categories.astype(int)
Out[10]: Int64Index([1, -9223372036854775808], dtype='int64')
```

dsaxton commented:
> That would be my preference. Did my earlier comment make sense?
> 
> The `NaN` issue can be solved by checking `Categorical.codes`
> 
> ```python
> In [7]: c = pd.Categorical([1, None, 2])
> 
> In [8]: c.codes < 0
> Out[8]: array([False,  True, False])
> ```
> 
> The inf issue is deeper, and is present in Float64Index, so I think it should be split to its own PR.
> 
> ```python
> In [9]: c = pd.Categorical([1, np.inf])
> 
> In [10]: c.categories.astype(int)
> Out[10]: Int64Index([1, -9223372036854775808], dtype='int64')
> ```

I believe so, you want to fix the `inf` casting behavior in one place (`Float64Index`) and then use the fixed `astype` method here?  Regarding the `nan` issue, is there a reason to prefer `self.codes < 0` over `self.isna()`?

Commit:
Only check NaN

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -520,6 +520,9 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
                 return self
             return self._set_dtype(dtype)
+        if is_integer_dtype(dtype) and self.isna().any():
+            msg = "Cannot cast to int."
```
```ValueError: cannot convert float NaN to integer``` is what we do now on
```pd.Series([1,2,np.nan],dtype='Int64').astype('int') ``` so would replicate this message

Commit:
Change error message

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -520,6 +520,9 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
                 return self
             return self._set_dtype(dtype)
+        if is_integer_dtype(dtype) and self.isna().any():
+            msg = "Cannot convert float NaN to integer"
```
Kinda pedantic, but we can have other NA values here.

```python
In [18]: cat = pd.Categorical([pd.Timestamp('2000'), pd.NaT])
```

In this case, it's not a float that we're refusing to cast. So perhaps `Cannot convert NA to integer`.

dsaxton replied:
Agree this seems more correct; do we want to think about consistency with `Series` as @jreback pointed out above?  Incidentally I just noticed that `Series` seems to be misbehaving as well in this special case, so probably worth a separate issue or PR:
```python
[ins] In [5]: pd.Series([pd.Timestamp("2000"), pd.NaT]).astype(int)             
Out[5]: 
0     946684800000000000
1   -9223372036854775808
dtype: int64
```

TomAugspurger commented on pandas/tests/extension/test_categorical.py:
```diff
@@ -197,7 +197,14 @@ def test_searchsorted(self, data_for_sorting):
 class TestCasting(base.BaseCastingTests):
-    pass
+    @pytest.mark.parametrize("cls", [Categorical, CategoricalIndex])
+    def test_cast_nan_to_int(self, cls):
```
Can you also parametrize this by datetime values like 

```
[pd.Timestamp('2000'), pd.NaT]
```

in addition to the numeric ones you already have?

Commit:
Add to test cases

jreback commented:
thanks @dsaxton
